[iOS] Admin Dashboard - Users

JPKribs opened the pull request:
### Reference
This is a rewrite of https://github.com/jellyfin/Swiftfin/pull/1278 with the goal of staying more focused on Users. I intend on adding password reset/changes later but for the sake of making this easier to review this is going to only tie back to users. Even that wasn't totally accomplished by this PR but it's much more focused than https://github.com/jellyfin/Swiftfin/pull/1278.

### Summary
Creates a list of all Users from the Jellyfin Server _(Mirroring DevicesView)_ and a DetailView. For now the DetailView is pretty empty but I linked it to the previous work done here: https://github.com/jellyfin/Swiftfin/pull/1277. This allows you to see devices owned only by a user. Additionally, this allows for deleting only the devices for a specific user when doing select all.

I have linked this UserView so it can be reached from the DeviceDetailsView by selecting the User or from the ActiveSessionsDetailView by selecting the User.

### Issues/Quirks
- My UserAdminViewModel has a loadDetails action. This is because it's possible to get to the user page with an incomplete UserDto. From DeviceDetailsView & ActiveSessionsDetailView you only have UserID and UserName so any future password resets or permission changes would not be possible. Loading Details ensures all information for that user exists in the Observer.
- There is a parameter in getDevices to filter by UserID. I don't know if this is an SDK issue or an API issue but this filter does not work. To get around this, I create var deviceBox which filters by the UserId. I want to eventually resolve the underlying issue but if it comes down to C# that's gonna bum me out...
- I use grayscale on user icons for users who are IsDisabled == True. Not a quirk but it could not the not-right way to go about it.

### Screenshots

<details>
    <summary>Updated Dashboard</summary>
    <img src="https://github.com/user-attachments/assets/e475b004-4fa4-4eaa-8694-0a53d1fc1770" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-23 at 15:51:20">
</details>

<details>
    <summary>User List</summary>
    <img src="https://github.com/user-attachments/assets/435085c5-98a2-4580-a810-1d7590436b37" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-23 at 15:51:28">
</details>

<details>
    <summary>User Details</summary>
    <img src="https://github.com/user-attachments/assets/1a7b15e3-8b09-4b18-b1cb-b10ae81f70f3" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-23 at 15:51:33">
</details>

<details>
    <summary>Filtered Devices</summary>
    <img src="https://github.com/user-attachments/assets/9fa715ee-0672-4317-a49a-023b64598218" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-23 at 15:51:55">
</details>

### Future Steps
**UsersView**
- Create New User (AddUserView)
- Delete User (**Discussion Required**)

**UserDetailsView**
- Password (Reset/Set New)
- Profile (UserDto.Configuration)
- Permissions (UserDto.Policy)
- Access (UserDto.Policy)
- Parental Controls (UserDto.Policy)

LePips commented on Shared/ViewModels/UserAdminViewModel.swift:
```diff
@@ -0,0 +1,232 @@
+import Foundation
+import JellyfinAPI
+
+final class UserAdminViewModel: ViewModel, Eventful, Stateful, Identifiable {
```
```suggestion
final class ServerUserAdminViewModel: ViewModel, Eventful, Stateful, Identifiable {
```

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/UsersView/Components/UsersRow.swift:
```diff
@@ -0,0 +1,162 @@
+        @CurrentDate
+        private var currentDate: Date
+
+        private var user: UserDto
```
```suggestion
        private let user: UserDto
```

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/UsersView/Components/UsersRow.swift:
```diff
@@ -0,0 +1,162 @@
+                        .lineLimit(2)
+                        .multilineTextAlignment(.leading)
+
+                    TextPairView(
```
I don't think this is that essential of information that it should be on every row. I do think it would be good to show something if the user is disabled though, but I have a different design in mind with a badge.

JPKribs replied:
I've removed this TextPair. Do you want me to take a shot at adding an icon for this or do you want me to leave this as is?

LePips replied:
I just tried my design and didn't entirely like it, I think I'm fine with just not showing that row.

JPKribs replied:
It looks like web just makes the disabled users grayscale so we're mirroring that functionality. An icon would be a nice to have but I think just the grayscale should be enough. I have that all on my latest commit.

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/UsersView/UsersView.swift:
```diff
@@ -0,0 +1,90 @@
+        .refreshable {
+            viewModel.send(.getUsers)
+        }
+        .topBarTrailing {
```
We may want to retrieve users based on the hidden/disabled filters and have a filter for that, if we want to do that now.

JPKribs replied:
Sorry, I don't think I fully understand what you mean. Do you mean add an action for getDisabledUsers / getHiddenUsers? Or add an extra parameter to the getUsers action like getUsers(active: true, inactive: false)?

LePips replied:
The latter. We can just have 2 `State` variables for the filter states and are set through a `Menu`.

JPKribs replied:
This is giving me some trouble since we have a state enum in the same struct. Is there a way to force this so it knows I mean the properly NOT the enum?
<img width="309" alt="Screenshot 2024-10-26 at 22 44 24" src="https://github.com/user-attachments/assets/bd38eceb-d214-4b88-900f-6ffbefd445ca">

LePips replied:
I am unsure of what exactly that is coming from. I mean having two `@State` variables within the view like below:

```swift
struct ... : View {

	@State
	private var isDisabledFilterActive = false
	@State
	private var isHiddenFilterActive = false
}
```

When these are updated, we update the device list.

JPKribs replied:
Ohhhh..... I misunderstood. I was trying to put states on the viewModel... Sorry about that! I know what direction to go now

JPKribs replied:
All right! I have this configured. This works how I would expect it and I have this defaulted to include all users regardless of hidden/disabled status. I think the only outlier now is making sure the filter is placed in a location that makes sense. I currently have it where it just sits on top of the user list.

LePips reviewed (changes requested):
Initial pass, looks great!

JPKribs commented on Swiftfin/Views/SettingsView/UserDashboardView/UsersView/Components/UsersRow.swift:
```diff
@@ -0,0 +1,162 @@
+                    .padding(.vertical, 8)
+            }
+            .onSelect(perform: onSelect)
+            // TODO: Do we even want to allow User Deletion?
```
I'm happy to add this but I'm kind of stuck again for the UI. I could do a delete & a multi-select for deletion but I also have a branch where I am working on allow user creation via an add button in the top left.

Would we want to have edit and add?

JPKribs replied:
I think I've got this! Let me know if this UI is bad but I think I've got a good starting point for this:
<img width="340" alt="Screenshot 2024-10-27 at 14 07 22" src="https://github.com/user-attachments/assets/0bbe0d2f-662d-4a74-b942-a1f08dd96cbb">
<img width="335" alt="Screenshot 2024-10-27 at 14 07 31" src="https://github.com/user-attachments/assets/043ca4c1-cde4-4ea6-bdaf-0025fe8176f1">

JPKribs replied:
I stole the Filters & Add button formatting from the SeeAllButton. I tried to get fancy with the icon on the outside but that might not be desirable.

As for the other select all/multi-select this is a pure drop in steal from how the DeviceView did this.

If we like this layout, after this PR I can add the filter option to the Devices as well. I also need to move Devices to be an array instead of using a box so I can do both of those at the same time. I don't even know what an add button would be on Devices so I think I would just add the filter and nothing else. For devices, I was thinking filter on Client?

LePips replied:
See: https://github.com/jellyfin/Swiftfin/pull/1287#issuecomment-2445014476

JPKribs commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/DevicesView.swift:
```diff
@@ -8,9 +8,12 @@
 import SwiftUI
 
 // TODO: Replace with CustomName when Available
+// TODO: Why doesn't this filter on the user from the Init?
```
The getUsers() call from the SDK is filtering or the filtering I added manually? That is, was this resulting in the filtered data Devices on only the selected user:

```
    init(userID: String? = nil) {
        _viewModel = StateObject(wrappedValue: DevicesViewModel(userID))
    }
```

I can give it another go, but I was getting all devices regardless of my userID

LePips replied:
Ah my bad. Looking at the server code, it looks like the intended use of the endpoint is not to filter on the user id, but to return all devices on the server that the user _could_ connect to, regardless if they ever have. It looks like the server doesn't do what we want it to do, so I may be less inclined to link to devices from users this way.

- https://github.com/jellyfin/jellyfin/blob/a0c634a6edcebdf0ee1ed28fc51dfefaab8b9392/Jellyfin.Server.Implementations/Devices/DeviceManager.cs#L193
- https://github.com/jellyfin/jellyfin/blob/a0c634a6edcebdf0ee1ed28fc51dfefaab8b9392/Jellyfin.Server.Implementations/Devices/DeviceManager.cs#L233

I think overall the device/user system on the server doesn't do what we want it to do since the device doesn't store all past users that have connected through it, but only the last one.

JPKribs replied:
That makes way more sense! I mean, we can still link through with 'lastUsername'. Primarily, what I see as valuable is the ability to nuke out all devices for a specific user. I wanted to add that as a toggle for a password reset function as well since this would help with a compromised password. What I have now gets us _pretty close_ to that functionality but it's definitely not 1 to 1 like I thought it was...

Does this still sound valuable or should I remove the User/Devices link?

LePips replied:
Let's just remove the User -> Devices link. When a user password is reset all old tokens are deleted, so we wouldn't need to nuke the devices to get that functionality.

JPKribs commented on Swiftfin/Views/SettingsView/UserDashboardView/UsersView/Components/UsersRow.swift:
```diff
@@ -0,0 +1,162 @@
+
+                        TextPairView(
+                            L10n.lastSeen,
+                            value: timeInterval <= twentyFourHours ?
```
Does this look like how you're expecting it to or should it _just_ be the date and not the time?
<img width="335" alt="Screenshot 2024-10-26 at 18 51 07" src="https://github.com/user-attachments/assets/ef5601b8-974f-41ed-82ec-bdf0b8d137b1">

LePips replied:
Just the date is good here. I can't remember what's done on the detail view, but we can present the time there since it's "more detail".

JPKribs commented:
I've got most of these changes in! I have a couple more design changes I'd love input on but I will pick this back up tomorrow/later in the week. I'm going to call it for now since I'm mostly happy with the current working version!

JPKribs commented:
I think I have all of the changes accounted for now. This PR is getting a tad bigger because I decided to move my addUser logic into this PR. I'm happy with how the UI looks but this ServerUsersView now allows:

- Creation
- Deletion (Bulk/Single)
- Filtering

So if this feels too clutter I won't be offended. I've left the 2 outstanding review changes that I'm unsure about open. One for design and one for the User/Device link.

JPKribs commented:
> I change the buttons to be in a `Menu`, as I don't really want to overload with buttons, but I may be too minimalistic. I've also made the delete button a pill button to live on the bottom toolbar and will want the other delete buttons to be migrated to it.
> 
> For user creation, that should live in its own view model.

I can take a look at this tomorrow! Thanks for the second set of eyes!

For the new viewModel, do you want a viewModel for the user list, user creation, user deletion or are we fine with a viewModel for the user list, & user deletion with the user creation on it's own viewModel only because it's on it's own View?

The reason I added user creation to the existing viewModel was to push the new user onto the .users set afterwards instead of reloading the whole list.

JPKribs commented:
> I change the buttons to be in a Menu, as I don't really want to overload with buttons, but I may be too minimalistic.

I had a chance to look this over. I'm actually a big fan! I think that might be a better look that some of the existing Edit/Add/Delete buttons I'm made for other parts of the Admin Dashboard. Do you want me to use this style for the other Edit/Add/Delete buttons in place of, for example, the add button so the ... is used there instead? Or should I leave views with only a single button as those single buttons?

I updated the Devices page to have the same Delete button.

I've updated everything related to users and I've tested that it's still all working as expected. I did see you commented out some of the backgroundStates.remove() calls in the ViewModels. I was guessing that's related to the warning they give since I'm 'not using' the results of the call. I'm not sure if there is a better way to do it but for now I have that re-enabled. Let me know if you have want to change this back.

Otherwise, I think this is ready to go from my end! Let me know if you want anything else adjusted!

LePips commented:
I did my cleanup and some refactoring for the device details since I noticed that was still in a box design and made its own view model.

For the add user view, that is now a modal since information creation/editing should now be in modals. Other design cleanup as well.

JPKribs commented:
> I did my cleanup and some refactoring for the device details since I noticed that was still in a box design and made its own view model.
> 
> For the add user view, that is now a modal since information creation/editing should now be in modals. Other design cleanup as well.

Thank you for your help cleaning this up! I do see 2 things I can address from my end but wanted your opinion:

Filtering users looks like this if there are no users (see below). I don't think this is bad but would we prefer this none is centered? Does the None need to have the separator removed? I'd guess this is the same style for the Devices as well but I don't think that matters since there will never be a case where there are no devices. Since, the device you are on will have to exist in the devices list.
![Simulator Screenshot - iPhone 16 Pro - 2024-10-31 at 13 40 01](https://github.com/user-attachments/assets/01617e00-09b0-43aa-8600-2ea2db8b2df6)

Secondly, when creating a user, I was having the viewModel for ServerUsersView always .getUsers onAppear to account for user creation. I don't think that was the *best* way to do this but the current version doesn't update the user list when a user is created. I see you have a comment for 

`TODO: somehow route to new user details`

on the addUser that kind of resolves this since the need to update the user list isn't there. That being said, for the time being do we feel fine using the onAppear() call for the ServerUserList for this purpose or is there a better route to add the new user to the ViewModel from another ViewModel?

LePips commented:
You can create a new notification in `SwiftfinNotifications` called `didAddServerUser` and use `onNotification` to listen when that is fired, passing in the new user object in the notification. From that, I would want us to push from the user list view to the user detail view so that the user could be further customized.

JPKribs commented:
> You can create a new notification in `SwiftfinNotifications` called `didAddServerUser` and use `onNotification` to listen when that is fired, passing in the new user object in the notification. From that, I would want us to push from the user list view to the user detail view so that the user could be further customized.

This was way easier than I thought it was going to be! I have something that is working how we expect it. All the existing Notification logic was really easy to follow and implement. I only have one quirk that works but might not be the best route. I put in a 0.1 second delay. Otherwise, the addServerUser `router.dismissCoordinator()` dismisses the new serverUserDetailsView that pops up. So, without the delay it all runs then immediately kicks me out of details. With this 0.1 second delay it triggers appropriately and at the end of the AddUser process, it leaves the user on the ServerUserDetailsView for the newly created UserDto.

Sorry to bother you further on this one. Let me know if there is anything I am missing or can improve on for this!

LePips commented:
At least for now with Stinsen, we can have an action fire after the coordinator is dismissed.

JPKribs commented:
> At least for now with Stinsen, we can have an action fire after the coordinator is dismissed.

Very cool! Thank you for your help on this!
